frontend/controller/actions/identity.js

```diff
@@ -701,5 +701,72 @@ export default (sbp('sbp/selectors/register', {
   ...encryptedAction('gi.actions/identity/joinDirectMessage', L('Failed to join a direct message.')),
   ...encryptedAction('gi.actions/identity/joinGroup', L('Failed to join a group.')),
   ...encryptedAction('gi.actions/identity/leaveGroup', L('Failed to leave a group.')),
-  ...encryptedAction('gi.actions/identity/setDirectMessageVisibility', L('Failed to set direct message visibility.'))
+  ...encryptedAction('gi.actions/identity/setDirectMessageVisibility', L('Failed to set direct message visibility.')),
+  'gi.actions/identity/uploadFiles': async ({ attachments, billableContractID }: {
+    attachments: Object, billableContractID: string
+  }) => {
+    const rootGetters = sbp('state/vuex/getters')
+
+    try {
+      const attachmentsData = await Promise.all(attachments.map(async (attachment) => {
+        const { mimeType, url } = attachment
+        // url here is an instance of URL.createObjectURL(), which needs to be converted to a 'Blob'
+        const attachmentBlob = await objectURLtoBlob(url)
+        const response = await sbp('chelonia/fileUpload', attachmentBlob, {
+          type: mimeType, cipher: 'aes256gcm'
+        }, { billableContractID })
+        const { delete: token, download: downloadData } = response
+        return {
+          downloadData: { ...omit(attachment, ['url']), downloadData },
+          deleteData: { token }
+        }
+      }))
+
+      const tokensByManifestCid = attachmentsData.map(({ downloadData, deleteData }) => ({
+        manifestCid: downloadData.downloadData.manifestCid,
+        token: deleteData.token
+      }))
+
+      await sbp('gi.actions/identity/saveFileDeleteToken', {
+        contractID: rootGetters.ourIdentityContractId,
+        data: { tokensByManifestCid }
+      })
+
+      return attachmentsData.map(({ downloadData }) => downloadData)
+    } catch (err) {
+      console.error('Error during uploading files', err)
+    }
+  },
+  'gi.actions/identity/removeFiles': async ({ manifestCids, identityContractID }: {
+    manifestCids: string[], identityContractID: string, billableContractID?: string
+  }) => {
+    const rootGetters = sbp('state/vuex/getters')
+    const me = sbp('state/vuex/state').loggedIn.identityContractID
+
+    if (me !== identityContractID) {
+      return
+    }
+
+    try {
+      const tokensMap = {}
+      for (const manifestCid of manifestCids) {
+        const token = rootGetters.currentIdentityState.fileDeleteTokens[manifestCid]
+        if (!token) {
+          console.error(`Missing delete token for file with manifestCid ${manifestCid}`)
+          return
```

Comment:
According to the @corrideat comment [here](https://github.com/okTurtles/group-income/pull/1892#discussion_r1557216751) and your comment [here](https://github.com/okTurtles/group-income/pull/1892#discussion_r1557926007), we can simply log warnings and continue process, can't we?

Reply:
@Silver-IT no, both of those comments are about behavior inside of contracts (after an action is sent, while it is being processed), whereas this is about the behavior of user-initiated actions (before they are successfully sent).

Reply:
`gi.actions/identity/uploadFiles` is the API that user can initiate. But, `gi.actions/identity/removeFiles` is not something that user can initiate. It's something that could be called in the `sideEffect` of `gi.contracts/chatroom/deleteMessage` and `gi.contracts/chatroom/deleteAttachments`.

Reply:
Ah, I see, in that case my comments do not apply to removeFiles, and if it makes sense to move that into a method inside of the contract, that might be a better place for it

Reply:
Actually, At first it was inside the contracts. But If we do that, we need to allow two sbp functions because they are being called inside the contract. That's why I made `gi.actions/identity/removeFiles`.
```
'chelonia/fileDelete'
'gi.actions/identity/removeFileDeleteToken'
```
 And another reason is to keep consistency. We have the similar function `gi.actions/identity/uploadFiles`, and I thought it's good to have `gi.actions/identityremoveFiles` which is much related to the former.

Reply:
Fair enough 👍
Maybe then just add a comment above it to mention that that action is intended to be called from a contract side effect, and not directly by the user